### ISSUE TITLE
Accessors is slow when type is parametric and not concrete in Julia <=1.10?

### ISSUE THREAD

**Issue by Tortar** [JuliaObjects/Accessors.jl#172]
MWE:

```julia
julia> using Accessors, BenchmarkTools

julia> struct A{X}
           a::X
           b
       end

julia> a = A{Int}(1, 1)
A{Int64}(1, 1)

julia> f1(a) = typeof(a)(a.a+1, a.b)
f1 (generic function with 1 method)

julia> f2(a) = @set a.a = a.a+1
f2 (generic function with 1 method)

julia> @benchmark f1($a)
BenchmarkTools.Trial: 10000 samples with 1000 evaluations.
 Range (min … max):  2.375 ns … 7.554 ns  ┊ GC (min … max): 0.00% … 0.00%
 Time  (median):     2.445 ns             ┊ GC (median):    0.00%
 Time  (mean ± σ):   2.469 ns ± 0.142 ns  ┊ GC (mean ± σ):  0.00% ± 0.00%

                    ▁   ▂   █                             ▇  
  ▂▁▁▂▁▁▁▄▁▁▁▆▁▁▆▇▁▁█▁▁▁█▁▁▁█▁▁▁▃▁▁▁▂▁▁▁▁▁▁▁▁▁▁▂▁▁▁█▁▁▄█▁▁█ ▃
  2.38 ns        Histogram: frequency by time       2.52 ns <

 Memory estimate: 0 bytes, allocs estimate: 0.

julia> @benchmark f2($a)
BenchmarkTools.Trial: 10000 samples with 10 evaluations.
 Range (min … max):  1.383 μs …  25.754 μs  ┊ GC (min … max): 0.00% … 0.00%
 Time  (median):     1.974 μs               ┊ GC (median):    0.00%
 Time  (mean ± σ):   1.991 μs ± 285.424 ns  ┊ GC (mean ± σ):  0.00% ± 0.00%

                          ▁▂▃▄▅▅▆█▆▇█▆▆▆▄▃▂▁                   
  ▂▂▁▂▁▂▂▂▁▂▂▁▂▂▂▂▂▃▃▃▅▅▆▇████████████████████▇▆▆▅▅▄▄▄▃▃▃▃▃▃▃ ▅
  1.38 μs         Histogram: frequency by time         2.4 μs <

 Memory estimate: 480 bytes, allocs estimate: 13.
```

shouldn't `f2` be as fast as `f1` ideally?

p.s. `Accessors` is amazingly useful!

**Comment by Tortar** [comment]
Seems like it looses the parameter type information:

```julia
julia> @code_warntype f2(a)
MethodInstance for f2(::A{Int64})
  from f2(a) @ Main REPL[5]:1
Arguments
  #self#::Core.Const(f2)
  a::A{Int64}
Body::A
1 ─ %1 = Core.apply_type(PropertyLens, :a)::Core.Const(PropertyLens{:a})
│   %2 = (%1)()::Core.Const((@o _.a))
│   %3 = (Accessors.opticcompose)(%2)::Core.Const((@o _.a))
│   %4 = (identity)(%3)::Core.Const((@o _.a))
│   %5 = Base.getproperty(a, :a)::Int64
│   %6 = (%5 + 1)::Int64
│   %7 = (Accessors.set)(a, %4, %6)::A
└──      return %7
```

**Comment by Tortar** [comment]
Are you sure you are on 1.10.5 and not on 1.11? Because I see your same good performance in 1.11 but not in 1.10.5, I updated to latest version to be sure, in 1.11 the code generation indeed doesn't lose the type information the parameters:

```julia
julia> @code_warntype f2(a)
MethodInstance for f2(::A{Int64})
  from f2(a) @ Main REPL[5]:1
Arguments
  #self#::Core.Const(f2)
  a::A{Int64}
Body::A{Int64}
1 ─ %1 = Core.apply_type(PropertyLens, :a)::Core.Const(PropertyLens{:a})
│   %2 = (%1)()::Core.Const((@o _.a))
│   %3 = (Accessors.opticcompose)(%2)::Core.Const((@o _.a))
│   %4 = (identity)(%3)::Core.Const((@o _.a))
│   %5 = Main.:+::Core.Const(+)
│   %6 = Base.getproperty(a, :a)::Int64
│   %7 = (%5)(%6, 1)::Int64
│   %8 = (Accessors.set)(a, %4, %7)::A{Int64}
└──      return %8
```

**Comment by Tortar** [comment]
oh yes, I actually had ConstructionBase 1.5.6 installed, now I'm on 1.5.8 and it is fast! I presume it was fixed in 1.5.7 with https://github.com/JuliaObjects/ConstructionBase.jl/pull/91, thanks for the help!
